Lib/tty.py

```diff
@@ -15,22 +23,136 @@
+    mode[LFLAG] &= ~(ECHO | ECHONL | ICANON | ISIG | IEXTEN)
+
+    # POSIX.1-2017, 11.1.7 Non-Canonical Mode Input Processing,
+    # Case B: MIN>0, TIME=0
+    # A pending read shall block until MIN (here 1) bytes are received,
+    # or a signal is received.
     mode[CC][VMIN] = 1
     mode[CC][VTIME] = 0
-    tcsetattr(fd, when, mode)
 
-def setcbreak(fd, when=TCSAFLUSH):
-    """Put terminal into a cbreak mode."""
-    mode = tcgetattr(fd)
-    mode[LFLAG] = mode[LFLAG] & ~(ECHO | ICANON)
+def cfmakecbreak(mode):
+    """cbreak mode termios"""
+    # Do not map CR to NL on input.
+    mode[IFLAG] &= ~(ICRNL)
+
+    # Do not echo characters; disable canonical input.
+    mode[LFLAG] &= ~(ECHO | ICANON)
+
+    # POSIX.1-2017, 11.1.7 Non-Canonical Mode Input Processing,
+    # Case B: MIN>0, TIME=0
+    # A pending read shall block until MIN (here 1) bytes are received,
+    # or a signal is received.
     mode[CC][VMIN] = 1
     mode[CC][VTIME] = 0
-    tcsetattr(fd, when, mode)
+
+def setraw(fd, when=TCSAFLUSH):
+    """Put terminal into raw mode.
+    Returns original termios."""
+    mode = tcgetattr(fd)
+    new = list(mode)
+    cfmakeraw(new)
+    tcsetattr(fd, when, new)
+    return mode
+
+def setcbreak(fd, when=TCSAFLUSH):
+    """Put terminal into cbreak mode.
+    Returns original termios."""
+    mode = tcgetattr(fd)
+    new = list(mode)
+    cfmakecbreak(new)
+    tcsetattr(fd, when, new)
+    return mode
+
+try:
+    from termios import TIOCGWINSZ, TIOCSWINSZ
+    HAVE_WINSZ = True
+
+    class winsize:
+
+        def __init__(self, row=0, col=0, xpixel=0, ypixel=0, fd=None):
+            """Creates winsize object.
+            If fd is not None, gets window size of tty of which fd is a
+            file descriptor. If fd is None, creates the winsize object
+            based on the row, col, xpixel, and ypixel values provided."""
+            if fd != None:
+                self.tcgetwinsize(fd)
+            else:
+                self.ws_row = row
+                self.ws_col = col
+                self.ws_xpixel = xpixel
+                self.ws_ypixel = ypixel
+
+        def __setattr__(self, name, value):
+            if not isinstance(value, int) or value < 0:
+                raise TypeError("expected nonnegative integer")
+            else:
+                super().__setattr__(name, value)
+
+        def __eq__(self, obj):
+            if isinstance(obj, self.__class__):
+                return self.__dict__ == obj.__dict__
+            else:
+                return False
+
+        def __repr__(self):
+            return f"winsize(ws_row={self.ws_row}, ws_col={self.ws_col}, ws_xpixel={self.ws_xpixel}, ws_ypixel={self.ws_ypixel})"
+
+        def __str__(self):
+            return f"(row={self.ws_row}, col={self.ws_col}, xpixel={self.ws_xpixel}, ypixel={self.ws_ypixel})"
+
+        def tcgetwinsize(self, fd):
+            """Gets window size of tty of which fd is a file descriptor."""
```

Comment:
@asvetlov Should tcgetwinsize() and tcsetwinsize() be defined in C as a part of termios.c [ the termios module ]? When tcgetwinsize() and tcsetwinsize() become standardized as a part of POSIX.1 issue 8, the C functions can be updated to use the native versions in termios.h; currently, they can simply use ioctl()+TIOCGWINSZ/TIOCSWINSZ.
If yes, then it would not be a bad idea to write cfmakeraw(), cfmakeecho(), and cfmakecbreak() etc as a part of termios.c either. Then tty.setraw() and tty.setcbreak() can simply make calls to those functions from the termios module. The "tty" library would then simply exist for backward compatibility; the original author of pty had created the tty module just to include the two helper functions tty.setraw() and tty.setcbreak(); it serves no other purpose.